data/scenarios/Challenges/_hackman/ghost.sw

```diff
@@ -0,0 +1,156 @@
+    turn left;
+    return isBlocked;
+    end;
+
+def chooseDirection : cmd dir =
+    leftBlocked <- checkLeftBlocked;
+    rightBlocked <- checkRightBlocked;
+    forwardBlocked <- isBlockedAhead;
+    if (leftBlocked && rightBlocked && forwardBlocked) {
+        say "Dead end; turning back";
+        return back;
+    } {
+        // Since we're in the "else" case of all three
+        // being blocked, we know that at least one
+        // direction is available. So we check all
+        // the combinations.
+        if (leftBlocked && rightBlocked) {
+            // Keep going straight
+            return forward;
+        } {
+            if (leftBlocked && forwardBlocked) {
+                return right;
+            } {
+                if (rightBlocked && forwardBlocked) {
+                    return left;
+                } {
+                    // Now we have exhaused all of the combinations
+                    // of *two* directions being blocked, so we
+                    // check single directions.
+
+                    if leftBlocked {
+                        // Decide whether to go right or straight
+                        d <- random 2;
+                        if (d == 0) {
+                            return right;
+                        } {
+                            // go straight
+                            return forward;
+                        }
+                    } {
+                        if rightBlocked {
+                            // Decide whether to go left or straight
+                            d <- random 2;
+                            if (d == 0) {
+                                return left;
+                            } {
+                                // go straight
+                                return forward;
+                            }
+                        } {
+                            if forwardBlocked {
+                                // Decide whether to go left or right
+                                d <- random 2;
+                                if (d == 0) {
+                                    return left;
+                                } {
+                                    // go right
+                                    return right;
+                                }
+                            } {
+                                // No directions are blocked, so we can
+                                // go left, right, or straight with
+                                // equal probability.
+
+                                d <- random 3;
+                                if (d == 0) {
+                                    return left;
+                                } {
+                                    if (d == 1) {
+                                        return right;
+                                    } {
+                                        // go straight
+                                        return forward;
```

Comment:
It would be really nice to have lists in the language, wouldn't it?  Then we could rewrite this tedious case analysis to just pick randomly from a list of available directions...